File: android/app/src/main/java/com/gooseco/technicalrulesquiz/ui/quiz/QuizViewModel.kt
package com.gooseco.technicalrulesquiz.ui.quiz

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.gooseco.technicalrulesquiz.data.model.AnsweredQuestion
import com.gooseco.technicalrulesquiz.data.model.QuizQuestion
import com.gooseco.technicalrulesquiz.data.model.QuizResult
import com.gooseco.technicalrulesquiz.data.repository.QuizRepository

class QuizViewModel(application: Application) : AndroidViewModel(application) {

    private val repository = QuizRepository(application.applicationContext)

    private val _currentQuestion = MutableLiveData<QuizQuestion>()
    val currentQuestion: LiveData<QuizQuestion> = _currentQuestion

    private val _currentQuestionNumber = MutableLiveData<Int>()
    val currentQuestionNumber: LiveData<Int> = _currentQuestionNumber

    private val _totalQuestions = MutableLiveData<Int>()
    val totalQuestions: LiveData<Int> = _totalQuestions

    private val _selectedAnswerIndex = MutableLiveData<Int?>()
    val selectedAnswerIndex: LiveData<Int?> = _selectedAnswerIndex

    private val _quizCompleted = MutableLiveData<Boolean>()
    val quizCompleted: LiveData<Boolean> = _quizCompleted

    private var questions: List<QuizQuestion> = emptyList()
    private var currentIndex = 0
    private val answeredQuestions = mutableListOf<AnsweredQuestion>()
    private var questionStartTime = 0L

    fun startQuiz(numberOfQuestions: Int = 10) {
        questions = repository.getRandomQuestions(numberOfQuestions)
        currentIndex = 0
        answeredQuestions.clear()
        _totalQuestions.value = questions.size
        _quizCompleted.value = false
        loadCurrentQuestion()
    }

    private fun loadCurrentQuestion() {
        if (currentIndex < questions.size) {
            _currentQuestion.value = questions[currentIndex]
            _currentQuestionNumber.value = currentIndex + 1
            _selectedAnswerIndex.value = null
            questionStartTime = System.currentTimeMillis()
        } else {
            _quizCompleted.value = true
        }
    }

    fun selectAnswer(answerIndex: Int) {
        _selectedAnswerIndex.value = answerIndex
    }

    fun submitAnswer() {
        val answerIndex = _selectedAnswerIndex.value ?: return
        val question = _currentQuestion.value ?: return

        val timeTaken = System.currentTimeMillis() - questionStartTime
        val isCorrect = answerIndex == question.correctAnswerIndex

        answeredQuestions.add(
            AnsweredQuestion(
                question = question,
                userAnswerIndex = answerIndex,
                isCorrect = isCorrect,
                timeTaken = timeTaken
            )
        )

        currentIndex++
        loadCurrentQuestion()
    }

    fun getQuizResult(): QuizResult {
        val correct = answeredQuestions.count { it.isCorrect }
        val incorrect = answeredQuestions.count { !it.isCorrect }

        return QuizResult(
            totalQuestions = questions.size,
            correctAnswers = correct,
            incorrectAnswers = incorrect,
            answeredQuestions = answeredQuestions
        )
    }

    fun hasSelectedAnswer(): Boolean {
        return _selectedAnswerIndex.value != null
    }
}

File: ios/PowerliftingQuiz/Sources/ViewModels/QuizViewModel.swift
import Foundation

@Observable
final class QuizViewModel {
    private(set) var currentQuestion: QuizQuestion?
    private(set) var currentQuestionNumber = 0
    private(set) var totalQuestions = 0
    private(set) var selectedAnswerIndex: Int?
    private(set) var quizCompleted = false
    private(set) var lastQuestionCount = 10

    private var questions: [QuizQuestion] = []
    private var currentIndex = 0
    private var answeredQuestions: [AnsweredQuestion] = []
    private var questionStartTime = Date()

    func startQuiz(count: Int, from repository: QuizRepository) {
        lastQuestionCount = count
        questions = repository.randomQuestions(count: count)
        currentIndex = 0
        answeredQuestions = []
        totalQuestions = questions.count
        quizCompleted = false
        selectedAnswerIndex = nil
        loadCurrentQuestion()
    }

    func selectAnswer(_ index: Int) {
        selectedAnswerIndex = index
    }

    func submitAnswer() {
        guard let index = selectedAnswerIndex,
              let question = currentQuestion else { return }

        answeredQuestions.append(AnsweredQuestion(
            question: question,
            userAnswerIndex: index,
            isCorrect: index == question.correctAnswerIndex,
            timeTaken: Date().timeIntervalSince(questionStartTime)
        ))

        currentIndex += 1
        loadCurrentQuestion()
    }

    func getQuizResult() -> QuizResult {
        let correct = answeredQuestions.filter(\.isCorrect).count
        return QuizResult(
            totalQuestions: questions.count,
            correctAnswers: correct,
            incorrectAnswers: questions.count - correct,
            answeredQuestions: answeredQuestions
        )
    }

    private func loadCurrentQuestion() {
        if currentIndex < questions.count {
            currentQuestion = questions[currentIndex]
            currentQuestionNumber = currentIndex + 1
            selectedAnswerIndex = nil
            questionStartTime = Date()
        } else {
            quizCompleted = true
        }
    }
}
